### DISCUSSION TITLE
Create Hive style directory hierarchy for Parquet via a Python program instead of using PARTITION_BY.

### DISCUSSION THREAD

**Discussion by parvizfarnia** [duckdb/duckdb#6384] [Q&A]
Hello everyone.

Since DuckDB v. 0.7.0 it is possible to create Parquet files with hive style, something like the following Python code sample:

`db_connection.execute("copy mytab to 'D:\\tmp\\parquet_data_dir' (format parquet, partition_by(A, B, C, D)))`

(We imagine that A, B, C and D are the name of columns in my table)

As you can see I partition based on 4 columns in my table which will be quite helpful for my end users who will import later the Parquet data based on different criteria relying on the values of these columns. However it seems, that partitioning with these four columns consumes a lot of memory, and each time after several minutes I get the following error message:

`duckdb.OutOfMemoryException: Out of Memory Error: could not allocate block of 262144 bytes`

Even when I set a temporary directory, something like `db_connection.execute("set temp_directory = 'D:\\parquet_tmp_dir'")` still the problem persists.

Today one of my colleagues suggested that I should create inside my program iteratively, inside multiple nested loops, the entire directory hierarchy and query the data based on the criteria of each directory via a `SELECT`. Then at each level call the `COPY` command in  DuckDB to create a single Parquet file for the data related to that level. In other words simulating the `PARTITION_BY` clause in the `COPY` command via nested loops.

Given that I consider myself a beginner both in DuckDB and Parquet, I preferred to ask the question here before developing to make sure that this makes sense. So basically here is the idea : 

```
# Obtaining as a list of tupples of all possible combinations of the directories
# that have to be created based on hive-style.
parquet_directories = db_connect.execute(
    "".join(
        [
            "select A, B, C, D from mytab ",
            "group by A, B, C, D ",
            "order by A asc, B asc, C asc, D asc"
        ]
    )
).fetchall() 

```

The result of this query is something like : 

```
| A  | B  | C  | D  |
======================
| a1 | b1 | c1 | d1 |
| a2 | b2 | c2 | d2 |
| a3 | b3 | c3 | d3 |
          .
          .
          .
```
          
          
These are indeed directories that have to be created based on hive-style while exporting the data based on Parquet format. For example, the first row in the above result set is indeed a directory that will have to be created with the following path

`A=a1/B=b1/C=c1/D=d1`

Using 4 nested loops (because I partition based on four columns), it is possible to create all of these directories. So each time if I put inside the 4th inner-most nested loop something like :

```
db_connect.execute(
    "".join(
        [
            "create table tmptab as  ",
            "select ... from mytab ",
            "where A='...' and B='...' and C='...' and D='...'
        ]
    )
)
```

And then in the same 4th inner-most nested loop I export this chunk of my data as a single Parquet file.

```
# Imagine the following will be updated inside my loops accordingly. The following
# value is just an example for the sake of this presentation
current_directory = "A=a1/B=b1/C=c1/D=d1"
#
#
db_connect.execute(
    "".join(
        [
            "copy tmptab to ",
            current_directory,
            "/data.parquet"
        ]
    )
)

And after that I DROP the tmptab for the next iteration in my loop.
```

Do you think that this will work and at the end, I will obtain the same parquet hierarchy, that I would have obtained if I had used `COPY` with `PARTITION_BY` clause ?

**Comment by parvizfarnia** [comment]
Just to be clear, the purpose of this thread is not asking for code, but only have the confirmation (yes or no) that splitting the process of creating a parquet multi-file dataset via `DuckDB` is possible based on the way that I explained above. In other words, in the case that `COPY` with `PARTITION_BY` cannot go to the end (because of lack of memory), then can it be said that the content of each of those leaves within the directory  tree which `PARTITION_BY` would have created, is completely independent and distinct from the others, in a way that in fact each of them could have been created as a standalon and single parquet file via `COPY` (without using `PARTITION_BY`)?

Thanks in advance.

**Reply by Mytherin** [reply]
Yes that is possible, a hive partitioned dataset is rather straightforward to manually create. Just make sure the partition is correct (i.e. parquet files in the folder `a=1/` all have `a=1`).

**Reply by parvizfarnia** [reply]
Thank you very much for your confirmation.
